velox/dwio/dwrf/reader/SelectiveDecimalColumnReader.cpp

```diff
@@ -87,56 +84,153 @@ void SelectiveDecimalColumnReader<DataT>::readHelper(RowSet rows) {
 
   // copy scales into scaleBuffer_
   ensureCapacity<int64_t>(scaleBuffer_, numValues_, &memoryPool_);
-  scaleBuffer_->setSize(numValues_ * sizeof(int64_t));
-  memcpy(
-      scaleBuffer_->asMutable<char>(),
-      rawValues_,
-      numValues_ * sizeof(int64_t));
+  size_t size = numValues_ * sizeof(int64_t);
+  scaleBuffer_->setSize(size);
+  memcpy(scaleBuffer_->asMutable<char>(), rawValues_, size);
 
-  // reset numValues_ before reading values
+  // reset numValues_ and outputRows_ before reading values
   numValues_ = 0;
+  outputRows_.clear();
   valueSize_ = sizeof(DataT);
   ensureValuesCapacity<DataT>(numRows);
 
   // decode value stream
   facebook::velox::dwio::common::
-      ColumnVisitor<DataT, common::AlwaysTrue, decltype(extractValues), kDense>
-          valueVisitor(filter, this, rows, extractValues);
+      ColumnVisitor<DataT, TFilter, ExtractValues, isDense>
+          valueVisitor(
+              *reinterpret_cast<TFilter*>(decodeFilter),
+              this,
+              rows,
+              extractValues);
   decodeWithVisitor<DirectDecoder<true>>(valueDecoder_.get(), valueVisitor);
-  readOffset_ += numRows;
+  auto scales = scaleBuffer_->as<int64_t>();
+  auto values = values_->asMutable<DataT>();
+
+  // if nullsPtr is NULL fillDecimals can go fast path
+  uint64_t* nullsPtr = nullptr;
+  if (decodeFilter->testNull()) {
+    VELOX_CHECK(
+        outputRows_.empty(),
+        "outputRows_ should be empty if decodeFilter accept Nulls.");
+    // if decodeFilter acceptNulls(e.g. AlwaysTrue), we should process all rows
+    for (auto row : rows) {
+      outputRows_.push_back(row);
+    }
+    // in this case nullsPtr is needed to judge whether the target value is NULL
+    if (nullsInReadRange_) {
+      nullsPtr = nullsInReadRange_->asMutable<uint64_t>();
+    }
+  }
+  DecimalUtil::fillDecimals<DataT>(
+      nullsPtr, values, values, scales, scale_, outputRows_, valuesFilter);
+
+  numValues_ = outputRows_.size();
+  rawValues_ = values_->asMutable<char>();
 }
 
 template <typename DataT>
 void SelectiveDecimalColumnReader<DataT>::read(
     vector_size_t offset,
     RowSet rows,
     const uint64_t* incomingNulls) {
-  VELOX_CHECK(!scanSpec_->filter());
+  velox::common::Filter* filter =
+      scanSpec_->filter() ? scanSpec_->filter() : &alwaysTrue();
+  if (filter->testNull() && version_ == velox::dwrf::RleVersion_2) {
+    /// RLEv2 does't support fastpath yet, temporarily disable bulkPathEnable_
+    /// to ensure that resultNulls_ will be allocated during prepareNulls()
+    /// or addNull() will be failed if the target Filter accept NULLs
+    bulkPathEnable_ = false;
+  }
   prepareRead<int64_t>(offset, rows, incomingNulls);
+  // enable fastpath after prepare has been finished
+  bulkPathEnable_ = true;
   bool isDense = rows.back() == rows.size() - 1;
-  if (isDense) {
-    readHelper<true>(rows);
+  if (scanSpec_->keepValues()) {
+    if (scanSpec_->valueHook()) {
```

Comment:
There is no aggregate pushdown defined for decimals.  Just `VELOX_CHECK(!scanSpec_->valueHook())` for now